samples/openapi3/client/petstore/python/tests_manual/test_api_call_headers.py

```diff
@@ -0,0 +1,90 @@
+# coding: utf-8
+
+"""
+    OpenAPI Petstore
+
+    This spec is mainly for testing Petstore server and contains fake endpoints, models. Please do not use this for any other purpose. Special characters: \" \\  # noqa: E501
+
+    The version of the OpenAPI document: 1.0.0
+    Generated by: https://openapi-generator.tech
+"""
+
+import sys
+from collections import namedtuple
+import json
+import unittest
+from unittest.mock import patch
+from urllib3.connection import HTTPHeaderDict
+
+import petstore_api
+from petstore_api.api.fake_api import FakeApi  # noqa: E501
+from petstore_api.rest import RESTResponse
+
+HTTPResponse = namedtuple(
+    'urllib3_response_HTTPResponse',
+    ['status', 'reason', 'data', 'getheaders', 'getheader']
+)
+
+headers = {'Content-Type': 'application/json'}
+def get_headers():
+    return {}
+def get_header(name, default=None):
+    return {}.get(name, default)
+
+
+class TestFakeApi(unittest.TestCase):
+    """FakeApi unit test stubs"""
+
+    def setUp(self):
+        self.api = FakeApi()  # noqa: E501
+
+    def tearDown(self):
+        pass
+
+    @staticmethod
+    def mock_response(body_value):
+        http_response = HTTPResponse(
+            status=200,
+            reason='OK',
+            data=json.dumps(body_value).encode('utf-8'),
+            getheaders=get_headers,
+            getheader=get_header
+        )
+        return RESTResponse(http_response)
+
+    def test_user_group_impersonation(self):
+        """Test case for passing multiple Impersonate-Group headers
+
+        """
+        from petstore_api.api_client import ApiClient
+        api_client = ApiClient()
+
+        with patch.object(api_client, 'request') as mock_method:
+            mock_method.return_value = self.mock_response([])
+
+            api_client.call_api(
+                '/fake',
+                'GET',
+                collection_formats = {'Impersonate-Group': 'multi'},
+                header_params = {
+                    'Impersonate-Group': ['dev', 'admin'],
+                },
+            )
```

Comment:
Hey there, this looks great. This is almost what I want. You correctly used response mocking and asserted that the mocked method was called with your expected inputs. But call_api should be a private method that is not used by a developer.
Can you call a real method that exists on the api_client and keep your assert called. with in line 74?
You can see examples of us doing that here: https://github.com/OpenAPITools/openapi-generator/blob/master/samples/openapi3/client/petstore/python/tests_manual/test_fake_api.py#L103

If you need to add a new endpoint to our python spec that you use for this test, you can add it to this file: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml

Reply:
Ah, well the reason I'm using `call_api` opens up a number of more questions about headers in the library. While the ApiClient has a `set_default_header` method, it has no mechanism to specify the collection format to use with the header.

The use of `collection_formats` in the code seems a bit odd in the way that this setting is used in so many different places in calls to `parameters_to_tuples` for headers as well as path, query, and post parameters... As best I can tell there is no reasonable way to configure the client with default `collection_formats` nor a way to pass a setting from a higher level call to the library... looking into another approach.

Reply:
Okay, I have pushed a new implementation and I think a better one. This moves the usage of HTTPHeaderDict to the default headers and adds a method to add header values, ex:

```
        api_client = ApiClient()
        api_client.add_default_header('Impersonate-Group', 'dev')
        api_client.add_default_header('Impersonate-Group', 'admin')
```

I still need to refine the test as you asked...